database/multistmt/parse.go

```diff
@@ -34,34 +36,65 @@ func splitWithDelimiter(delimiter []byte) func(d []byte, atEOF bool) (int, []byt
 // Parse parses the given multi-statement migration
 func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, h Handler) error {
 	// notes:
-	// 1. comment chars will be detected anywhere, a '--' in the middle of a 
+	// 1. comment chars will be detected anywhere, a '--' in the middle of a
 	//    line will start comment mode(good and bad)
-	// 2. input can be arbitrarily large, but the internal buffers will be 
+	// 2. input can be arbitrarily large, but the internal buffers will be
 	//    problems(like statements)
 	// 3. could be converted to work with logger, for now fmt is still used
 	// 4. doesn't support /* */ c-style comments (future)
 	// 5. doesn't support nested comments (future)
+	// 6. now supports plpgsql trigger bodies
 	var err error = nil
 	// buf is the bytes read from input reader
 	buf := make([]byte, 1024)
 	// true when we're ignoring input(during comments)
 	discard := false
+	// fnbody is true when a function body delimiters $$ are encountered
+	fnbody := false
 	// accumulate statements intermediate buffer, this buffer will be incomplete
 	// until end-of-statement char ';'
 	accum := make([]byte, 0, 1024)
 	// completed statements, contents of accum will be dumped in here
 	stmts := make([][]byte, 0, 1000)
+
 	for err == nil {
 		n, err := reader.Read(buf)
 		if n > 0 {
 			for i := range buf[:n] {
-				// when first two chars are comment indicators.
-				if len(buf) > 1 &&  buf[i] == '-' && buf[i+1] == '-' {
-					discard = true
+				if !fnbody {
+					// when first two chars are comment indicators.
+					switch {
+					// ignore all lines that start with --
+					case len(buf) > 1 && i+1 < len(buf) && buf[i] == '-' && buf[i+1] == '-':
+						discard = true
+					// ignore any lines that start with // (this also covers ///)
+					case len(buf) > 1 && i+1 < len(buf) && buf[i] == '/' && buf[i+1] == '/':
+						discard = true
+					}
 				}
+
+				// output the content, for logging
 				fmt.Printf("%c", buf[i])
 				switch ch := buf[i]; ch {
+				case '$':
+					// look around is there another $?
+					// is there also and ending marker like "$$ LANGUAGE plpgsql"
+					if buf[i+1] == '$' {
+						// set fnbody false to trigger the check for the next `;`
+						if fnbody {
+							fnbody = false
+						} else {
+							fnbody = true
+						}
```

Comment:
```suggestion
						fnbody = !fnbody
```

Simpler, but perhaps debatable if it's more readable.